File: tools/terrain-editor/src/heightmap.cpp
#include "heightmap.h"
#include <fstream>

static constexpr uint32_t kMagic = 0x4D484352; // "RCHM"

bool Heightmap::Save(const std::string& path) const {
    std::ofstream f(path, std::ios::binary);
    if (!f) return false;
    f.write(reinterpret_cast<const char*>(&kMagic),     4);
    f.write(reinterpret_cast<const char*>(&W),          4);
    f.write(reinterpret_cast<const char*>(&H),          4);
    f.write(reinterpret_cast<const char*>(&cell_size),  4);
    f.write(reinterpret_cast<const char*>(heights.data()), heights.size() * 4);
    return true;
}

bool Heightmap::Load(const std::string& path) {
    std::ifstream f(path, std::ios::binary);
    if (!f) return false;
    uint32_t magic = 0;
    f.read(reinterpret_cast<char*>(&magic), 4);
    if (magic != kMagic) return false;
    f.read(reinterpret_cast<char*>(&W),         4);
    f.read(reinterpret_cast<char*>(&H),         4);
    f.read(reinterpret_cast<char*>(&cell_size), 4);
    heights.resize(W * H);
    f.read(reinterpret_cast<char*>(heights.data()), heights.size() * 4);
    return !!f;
}

File: server/internal/world/heightmap.go
package world

import (
	"encoding/binary"
	"fmt"
	"math"
	"os"
)

const heightmapMagic uint32 = 0x4D484352 // "RCHM"

type Heightmap struct {
	w, h     int
	cellSize float32
	data     []float32
}

func LoadHeightmap(path string) (*Heightmap, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer f.Close()

	var magic, w, h uint32
	var cellSize float32
	if err := binary.Read(f, binary.LittleEndian, &magic); err != nil {
		return nil, err
	}
	if magic != heightmapMagic {
		return nil, fmt.Errorf("heightmap: bad magic %08X", magic)
	}
	binary.Read(f, binary.LittleEndian, &w)
	binary.Read(f, binary.LittleEndian, &h)
	binary.Read(f, binary.LittleEndian, &cellSize)

	data := make([]float32, w*h)
	if err := binary.Read(f, binary.LittleEndian, data); err != nil {
		return nil, err
	}
	return &Heightmap{w: int(w), h: int(h), cellSize: cellSize, data: data}, nil
}

func (hm *Heightmap) SampleWorld(x, z float32) float32 {
	fx := x / hm.cellSize
	fz := z / hm.cellSize
	ix := int(math.Floor(float64(fx)))
	iz := int(math.Floor(float64(fz)))
	tx := fx - float32(ix)
	tz := fz - float32(iz)
	h00 := hm.get(ix, iz)
	h10 := hm.get(ix+1, iz)
	h01 := hm.get(ix, iz+1)
	h11 := hm.get(ix+1, iz+1)
	return lerp(lerp(h00, h10, tx), lerp(h01, h11, tx), tz)
}

func (hm *Heightmap) get(x, z int) float32 {
	if x < 0 {
		x = 0
	} else if x >= hm.w {
		x = hm.w - 1
	}
	if z < 0 {
		z = 0
	} else if z >= hm.h {
		z = hm.h - 1
	}
	return hm.data[z*hm.w+x]
}

func lerp(a, b, t float32) float32 { return a + (b-a)*t }
